### ISSUE TITLE
 error_chain's Result<T> support

### ISSUE THREAD

**Issue by Alex-PK** [intellij-rust/intellij-rust#1243]
When using the error_chain crate it's common to define the errors in a "errors.rs" mod and import them where needed with `use errors::*`.

This overrides the default `Result<T,E>` enum with a `Result<T>`, and intellij-rust reports all the functions using it as errors (Wrong number of arguments: expected 2, found 1 [E0243]).

I know it's hard to parse macros (see #955), but I was wondering if there is a way to use annotations to declare the definition of enums, structs or functions (like `// @defines Result<T>` or similar) to use in the mod defining them, so that the parser recognizes them.

P.S.: sorry for the empty bug report earlier, I don't know what I pressed to post it while writing... :)

**Comment by alygin** [comment]
Yeah, macros are a sore point for the plugin's code analysis machinery at the time. I see two possible solutions here:
1. Make E0243 an inspection instead of annotation, so it could be opted out.
2. Introduce a special case for the `error_chain!` macro. At least for its simplest usage, without the `types` block (and expand it later).

@matklad, what do you think? Any other ideas?

**Comment by matklad** [comment]
>Make E0243 an inspection instead of annotation, so it could be opted out.

Yeah, we definitely should do this with perhaps most of the errors: combination of macros and shadowing leads to false positives :( 

Special casing `error_chain!` would be a nice addition!

**Comment by alygin** [comment]
> Special casing error_chain! would be a nice addition!

Are there examples of such macro expansions in the plugin? I saw special cases for several macros, but they seem to only handle macro calls correctly, not to expand them.

**Comment by matklad** [comment]
>Are there examples of such macro expansions in the plugin? I saw special cases for several macros, but they seem to only handle macro calls correctly, not to expand them.

No, there are no such examples yet. The general idea would be to create an RsFile with macro expanded code (the same way `RsPsiFactory` creates files), find the relevant nodes in this file and attach them via `getContext` mechanism to the macro call. We now handle name resolution of strings like `foo::bar::baz` in the similar way (the place to look at is `RsCodeFragmentFactory`).

So in the end you'll have a `MacroCall` node, which has an `.expansion` property which holds a PSI node in an in-memory file. During resolve of a name, when you encounter a `MacroCall`, you process `.expansion`. And when you are resolving a name *inside* the expansion, everything should just work because resolve uses `.getContext` and not `.getParent`.

And there should be macro hygene somewhere in here, but we can definitely punt on it for the start :) 

Does this make any sense?

**Comment by alygin** [comment]
It surely makes sense, though will probably take some time to digest and implement. But that looks like another small step towards the general macro expansion mechanism, so it's worth doing.

I'll start with extracting the most false-positvie-prone errors into inspections and then have a look at macro expansion.

**Comment by rbalicki2** [comment]
Before checking here, I also made a stack overflow post on this btw http://stackoverflow.com/questions/44089097/rust-ide-not-detecting-result-from-error-chain-thinks-im-using-the-stdresult

**Comment by alygin** [comment]
This error annotation is an inspection now (along with other similar), so you can disable it entirely in the project preferences or suppress for the particular function.

**Comment by josephDunne** [comment]
Sorry, thanks for creating the opt out but what is the correct way to handle this going forward? At last count my project has 200+ Result<T> return types spread over all my modules and I don't want to disable wrong type parameter number for my whole project

**Comment by matklad** [comment]
>Sorry, thanks for creating the opt out but what is the correct way to handle this going forward?

The correct way would be to expand error-chain macros. https://github.com/intellij-rust/intellij-rust/issues/1243#issuecomment-302644658 is a good start in that direction.

**Comment by Alex-PK** [comment]
Sorry, but I agree with @josephDunne here. I am grateful for the inspection option, but I don't think it fixes this issue. I can't disable it completely, as it will be useful in other contexts, and setting it as a warning doesn't solve the problem.

**Comment by matklad** [comment]
Yeah, let's reopen it!

**Comment by tarcieri** [comment]
@matklad is there a tracking issue for supporting macro expansions in general?

**Comment by avkonst** [comment]
I type errors::Result<> everywhere instead of Result<, but it is useless as the caller function does not know the return type anyway (it is reported as '<unknown>') and further intellisense is broken..

**Comment by avkonst** [comment]
I wonder if there are any plans on progressing this issue and when?

**Comment by FauxFaux** [comment]
I've been using this as a workaround, if anyone else is interested:

https://github.com/FauxFaux/error-chain-for-dumb-ides#error-chain-for-dumb-ides

**Comment by avkonst** [comment]
Excellent. It is like headers for closed source in c++ or typescript. It is
useful also for humans. Viewing source of macros or declaration does not
tell as much than these sort of stubs! Maybe it should be a good practice
for rust developers to supply stubs for macros?

**Comment by vlad20012** [comment]
Fixed with #2331
